### PR TITLE
Auto (re-)connection

### PR DESCRIPTION
**Pull request by ghost**
This adds a new BackgroundConnect option, which allows a caller of
Connect() to avoid thinking about database availabilty at startup and
while running.

This means the caller never sees connection errors at startup time if
the neo4j instance in unavailable. Instead the calling application can
start up, but all database operations will fail while there is no connection.

If an underlying error is detected while running that indicates a reconnection
is required, a new connection is created in the background.

### CODE REVIEW

File: neoutils/connection.go [hunk 1]
```diff
@@ -37,6 +41,18 @@ func Connect(neoURL string, conf *ConnectionConfig) (NeoConnection, error) {
 		conf = DefaultConnectionConfig()
 	}
 
+	if !conf.BackgroundConnect {
+		return connectDefault(neoURL, conf)
+	} else {
+		f := func() (NeoConnection, error) {
+			return connectDefault(neoURL, conf)
+		}
+		return connectAuto(neoURL, f, 30*time.Second)
```
Comment (ghost) [inline neoutils/connection.go:54]: Is the 30 seconds a first guess, or a calculated amount of time to wait between connection attempts?

Comment (ghost) [reply neoutils/connection.go:54]: It was pretty randomly chosen. Too short means causing noise and load.  Too long means more downtime than needed.

We can easily tweak in future, adding a config option if that makes sense.

File: neoutils/auto_connect_test.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+	conn, err := connectAuto("http://localhost:9999/db/data/", connect, period)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	<-triedConnect
+	// should not connect again
+	select {
+	case <-triedConnect:
+		t.Fatal("shoulu not have reconnected")
```
Comment (tamas-molnar) [inline neoutils/auto_connect_test.go]: Typo, appears a few times in this function.

**Comment by tamas-molnar** [comment]
Nice, I especially like the tests :+1:

**Comment by carlosroman** [comment]
Tested it locally, it does exactly what it says on the tin 👍

**Pull request merged**